packages/protocol/contracts/bridge/libs/LibBridgeProcess.sol

```diff
@@ -73,21 +77,40 @@ library LibBridgeProcess {
             revert B_STATUS_MISMATCH();
         }
 
-        // Check if the signal has been received on the source chain
-        address srcBridge =
-            resolver.resolve(message.srcChainId, "bridge", false);
+        if (checkProof) {
+            // Check a chain of inclusion proofs, from the message's source
+            // chain all the way to the destination chain.
+            uint256 srcChainId = message.srcChainId;
+            address app = resolver.resolve(message.srcChainId, "bridge", false);
+            bytes32 signal = msgHash;
+
+            bool verified;
+
+            for (uint256 i; i < proofs.length - 1; ++i) {
+                IBridge.IntermediateProof memory iproof =
+                    abi.decode(proofs[i], (IBridge.IntermediateProof));
+                // perform inclusion check
+                verified = LibSecureMerkleTrie.verifyInclusionProof(
+                    bytes.concat(LibSignalService.getSignalSlot(app, signal)),
+                    hex"01",
+                    iproof.mkproof,
+                    iproof.signalRoot
```

Comment:
I think supplying `iproof.signalRoot` with the arguemnts does not guarantee that the proof is not 'made-up' (tamper proof).

In current version, how we call the verifyInclusionProof is like that:
![kép](https://github.com/taikoxyz/taiko-mono/assets/51912515/bdf83821-dca2-4f43-b083-e06d3660acc3)

So we know the signalRoot can be trusted because it is synced with the anchor, no ? 🤔

Reply:
Maybe it is overkill and there are easier solutions, my .02:

To have a trusted `signalRoot`, so with an example `L2B` have a trusted , synched `L2A` and `L2C` signal root, i drafted a quick design.

So the L1 <-> L2 would not been touched !


BUT in case the direction is not like that (rather L2 <-> L2), we could have a registry contract which only out special relayer could update. See diagram:

![kép](https://github.com/taikoxyz/taiko-mono/assets/51912515/fc4ecf75-c242-423c-89bf-bb07fc278fdc)

Now, when someone wants to bridge from `L2A` to `L2B` 2 proof needed:

1. First proof: the registry synced signal root confirm inclusion (of message)
2. Second proof: The L1 relayed (with anchor), common SignalService contract contains L2A SignalRoot as a storage slot with value: 1

In the UI / Client + also in the contract:
- If it is a L1 <-> L2 operation proof.length = 1
- Otherwise validate the first proof based on this `'Taiko Deployed Registry'` and the second shall be that signalRoot is within the L1 common relayed one.